packages/binding-coap/src/coap-server.ts

```diff
@@ -337,387 +335,471 @@ export default class CoapServer implements ProtocolServer {
+                } else {
+                    this.handleObserveProperty(req, thing, res, affordanceKey);
+                }
+                break;
+            case "PUT":
+                if (property.readOnly === true) {
+                    this.sendErrorResponse(res, "4.00", "Property readOnly");
+                    return;
+                }
+
+                this.handleWriteProperty(property, req, contentType, thing, res, affordanceKey);
+                break;
+            default:
+                this.sendMethodNotAllowedResponse(res);
+        }
+    }
+
+    private async handleReadProperty(
+        property: PropertyElement,
+        req: IncomingMessage,
+        contentType: string,
+        thing: ExposedThing,
+        res: OutgoingMessage,
+        affordanceKey: string
+    ) {
+        try {
+            const interactionOptions = this.createInteractionOptions(
+                property.forms,
+                property.uriVariables,
+                thing,
+                req,
+                contentType
             );
-        });
+            const content = await thing.handleReadProperty(affordanceKey, interactionOptions);
+            this.streamContentResponse(res, content);
+        } catch (err) {
+            error(`CoapServer on port ${this.getPort()} got internal error on read '${req.url}': ${err.message}`);
+            this.sendErrorResponse(res, "5.00", err.message);
+        }
+    }
 
-        const requestUri = req.url;
-        let contentType = req.headers["Content-Format"] as string;
+    private async handleObserveProperty(
+        req: IncomingMessage,
+        thing: ExposedThing,
+        res: OutgoingMessage,
+        affordanceKey: string
+    ) {
+        const listener = async (content: Content) => {
+            try {
+                this.streamContentResponse(res, content, { end: true });
+            } catch (err) {
+                error(`CoapServer on port ${this.getPort()} got internal error on read '${req.url}': ${err.message}`);
+                this.sendErrorResponse(res, "5.00", err.message);
+            }
+        };
 
-        if (req.method === "PUT" || req.method === "POST") {
-            if (!contentType && req.payload) {
-                warn(
-                    `CoapServer on port ${this.getPort()} received no Content-Format from ${Helpers.toUriLiteral(
-                        req.rsinfo.address
-                    )}:${req.rsinfo.port}`
-                );
-                contentType = ContentSerdes.DEFAULT;
-            } else if (
-                ContentSerdes.get().getSupportedMediaTypes().indexOf(ContentSerdes.getMediaType(contentType)) < 0
-            ) {
-                res.code = "4.15";
-                res.end("Unsupported Media Type");
-                return;
+        thing
+            .handleObserveProperty(affordanceKey, listener, null)
+            .then(() => res.end())
+            .catch(() => res.end());
+
```

Comment:
Sorry for the current formatting is hard to tell, but can't we use `trycatch` and `await` here?

Reply:
Good point, that has been a part of the original implementation which I found kind of strange therefore I didn't want to touch it initially 😅 I now tried to provide a proper solution as you proposed in 3b8604c4eb20954802c338b363e964f40956796e. Not sure if always invoking `res.end()` as in the original code is the correct approach here, though.

Reply:
Slightly adjusted the code in 0bf0d5af70fbfbefcac2bc238084574a74b7cf6a to be closer to the original logic.